README.txt
Pyjsdl - Python-to-JavaScript Multimedia Framework

Pyjsdl module is modelled on Pygame/SDL commands that wraps JavaScript functionality including HTML5 canvas. The module permits scripts coded in Python/Pygame to compile to JavaScript using the Pyjs compiler (http://pyjs.org), allowing deployment of JavaScript applications without extensive editing of the script. The current version of the module supports a subset of Pygame functionality. Information concerning use of the Pyjsdl module is included in guide.txt.

Pyjsdl is released under the MIT License, see LICENSE.txt for further information.

Pyjsdl page: http://gatc.ca/projects/pyjsdl/
Pyjsdl docs: http://gatc.ca/projects/pyjsdl/doc/



history.txt
Pyjsdl Version History

0.10    April 24, 2013
-initial release

0.11    April 26, 2013
-updated Sprite.clear to allow callable argument.
-updated Surface._blit_clear with clipping.
-updated Surface.blits to directly call drawImage.
-updated Surface.fill rect argument.
-updated Draw.arc, Transform.rotate, and Rect with int() division.

0.12    April 29, 2013
-added mouse events to Canvas.
-updated Event to process JS mouse events.
-updated Mouse.get_pressed and Mouse.get_pos for JS mouse events.
-updated Time.wait to use JS Timeout.

0.13    May 4, 2013
-added key events to Canvas.
-updated Event for process JS key events.
-updated Mouse methods variable.

0.14    May 15, 2013
-added Font rendering to canvas.
-updated Event so key event works in IE browser.
-updated Display.update rect argument for pyjs -O compilation.
-updated Event key handling for IE browser.
-updated Sprite Group __contains__ to work with pyjs.
-updated Surface.blit to take rect argument and return rect.
-updated Surface.fill color argument.
-modified pyjs0.8+ git-130513build (pyjs_changes.zip) i. IE9+ compatibility: edited pyjs/src/pyjs/boilerplate/home.nocache.html and pyjs/src/pyjs/browser.py, and added <!DOCTYPE html> to top of main program html in pyjs output. ii. HTML5canvas usage: renamed library/pyjamas/Canvas/HTML5Canvas.ie6.py.

0.15    July 30, 2013
-added Mask using Pyjsarray and Pyjsbitset.
-added Pyjsarray and Pyjsbitset based on JavaScript TypedArray object.
-added Surface get_at/set_at using Canvas getImageData and putImageData.
-updated Surface.subarray.
-updated Canvas and Surface to use pyjamas.Canvas.HTML5Canvas.
-updated Rect.createIntersection.
-updated pyjs_changes.zip - Pyjs git updated for HTML5canvas usage.

0.16    September 1, 2013
-added Surfarray using Pyjsarray.
-updated Pyjsarray with Ndarray and ImageData objects.
-updated image.load
-updated display.update rect processing
-added display.update_rect method
-updated display.update/display.update_rect/surface._blit_clear rect clipping

0.17    November 1, 2013
-updated Pyjsarray ImageData object for IE compatibility.
-added util.call to call unbound methods.
-updated Surface.blit to int convert argument.
-updated Rect.inflate and Rect.inflate_ip for proper positioning.
-added Textbox/Textarea and display.textbox_init for text input.
-updated mouse positioning with page scroll position.
-updated Surfarray.blit_array.
-added Pyjsarray Ndarray methods.

0.18    February 27, 2014
-updated Pyjsarray typedarray constructor for Chrome compatibility.
-updated Surface constructor to int convert argument.
-updated Transform.rotate positioning.
-updated Transform rotate/flip to restore context.
-added Textbox/Textarea resize method.
-updated Surface.resize for width/height attributes.
-updated Canvas.resize to align elements.
-updated Canvas to subclass Surface.
-updated Canvas with buffered surface optional.
-added display.is_canvas and updated Canvas blit not buffered/image load not converted to canvas surface in IE8; Pyjs VML-Canvas requires IE8 quirk mode - no <!DOCTYPE html>.
-updated Surface constructor to take optional arguments.
-updated Event.poll noevent return.
-updated Surface.get_at to return color tuple.
-updated Rect attribute positioning.
-added Rect equality and nonzero methods.
-added util.Pyjs_Mode to check Pyjs compilation mode.
-added Color object.
-updated to use Color object.
-updated Rect constructor.
-updated draw methods with minor adjustments.
-updated Mask.fill method.
-updated to use Rect object.
-updated Sprite group update to account for member changes.
-updated util.Timer for output to textarea rather than console logger.
-updated Surfarray array2d PyImageInteger processing.
-updated Surfarray make_surface method.
-added draw.ellipse method.
-updated Pyjsarray Typedarray constructor when type not implemented.
-changed license from GPL to LGPL.

0.19    August 20, 2014
-updated bitset array to avoid js reserved work.
-updated quit function to terminate program iteration.
-updated transform.rotozoom for correct scaling.
-add env.pyjs_mode with strict/optimized bool attributes to check pyjs-S/-O mode.
-correct modifier keys keyevent detect with pyjs -S compilation (worked in -O) with sets membership testing of onKeyDown keycode that appears due to js/pyjs numeric difference.
-update display event to properly clear modifier keys held with onMouseLeave event.
-update display set_caption/get_caption to access Canvas element id.
-add display.get_active method.
-restructure event handler to isolate from UserEvent and JEvent objects.
-update display.update for performance.
-update event eventtype list as a set object.
-update rect inflate/inflate_ip positioning.

0.20    November 9, 2014
-add rect union methods.
-update methods to use rect intersection.
-correct surface subsurface rect access.
-add rect rectPool to utilize a rect pool.
-update sprite draw for performance.
-update sprite collide methods processing.
-add rect clamp methods.
-update display setup callback function.
-update display repaint process.
-add mixer.

0.21_git
-update font to improve fonts access.
-add rect.contains method.
-update image load methods for imagedata objects.
-add display.setup_images method.
-add key.name method.
-add mask.toString method.
-update display.update to properly handle clipping.
-update surface.blit with clipping.
-update rect union methods to process properly.
-add rect iter method.
-add rect collide methods - collidelistall, collidedict, and collidedictall.
-add sprite collide methods - collide_rect_ratio, collide_circle, and collide_circle_ratio.
-refactoring to isolate pyjs.
-refactoring for optimization.
-correct event.pump to maintain queue.
-update IE9+ compatibility files in pyjs_changes.zip.
-add mousewheel event listener.



font.py
#Pyjsdl - Copyright (C) 2013 James Garnon <http://gatc.ca/>
#Released under the MIT License <http://opensource.org/licenses/MIT>

#from __future__ import division
from surface import Surface
import math
from color import Color
from pyjsobj import HTML5Canvas

__docformat__ = 'restructuredtext'


_initialized = False
_surf = None


def init():
    """
    **pyjsdl.font.init**
    
    Initialize font module.
    """
    global _surf, _initialized, match_font
    if _initialized:
        return
    try:
        _surf = HTML5Canvas(1,1)
        _surf.measureText('x')
    except:
        _surf = None
    _initialized = True
init()


def quit():
    """
    **pyjsdl.font.quit**
    
    Unintialize font module.
    """
    global _surf, _initialized
    _surf = None
    _initialized = False


def get_init():
    """
    **pyjsdl.font.get_init**
    
    Check if font module is initialized.
    """
    return _initialized


def get_default_font():
    """
    **pyjs2d.font.get_default_font**
    
    Return default font.
    """
    return Font._font[0]


def get_fonts():
    """
    **pyjsdl.font.get_fonts**
    
    Return font names, which have fallback fonts if unavailable.
    """
    return Font._font


def match_font(name):
    """
    **pyjsdl.font.match_font**
    
    Argument name is a font name, or comma-delimited string of font names.
    Return font string in compliant format.
    """
    fallback = False
    font = [fn.strip().lower() for fn in name.split(',')]
    for i, fn in enumerate(font):
        if fn in Font._font:
            fallback = True
            continue
        else:
            f = ''.join(c for c in fn if c.isalnum())
            if f in Font._font_alt:
                font[i] = Font._font[Font._font_alt[f]]
                fallback = True
    if not fallback:
        font.append(Font._font[0])
    font = ','.join(font)
    return font


class Font(object):
    """
    **pyjsdl.font.Font**
    
    * Font.render
    * Font.size
    * Font.set_underline
    * Font.get_underline
    * Font.set_bold
    * Font.get_bold
    * Font.set_italic
    * Font.get_italic
    * Font.get_linesize
    """

    _font = ['arial', 'bitstream vera sans', 'bitstream vera serif', 'book antiqua', 'comic sans ms', 'courier new', 'courier', 'dejavu sans', 'dejavu sans mono', 'dejavu serif', 'freesans', 'garamond', 'georgia', 'helvetica', 'impact', 'liberation sans', 'liberation serif', 'lucida console', 'lucida serif', 'nimbus mono l', 'nimbus roman no9 l', 'nimbus sans l', 'palatino', 'times new roman', 'times', 'tahoma', 'verdana', 'cursive', 'monospace', 'sans-serif', 'serif']

    _font_alt = {'arial':0, 'bitstreamverasans':1, 'bitstreamveraserif':2, 'bookantiqua':3, 'comicsansms':4, 'couriernew':5, 'courier':6, 'dejavusans':7, 'dejavusansmono':8, 'dejavuserif':9, 'freesans':10, 'garamond':11, 'georgia':12, 'helvetica':13, 'impact':14, 'liberationsans':15, 'liberationserif':16, 'lucidaconsole':17, 'lucidaserif':18, 'nimbusmonol':19, 'nimbusromanno9l':20, 'nimbussansl':21, 'palatino':22, 'timesnewroman':23, 'times':24, 'tahoma':25, 'verdana':26, 'cursive':27, 'monospace':28, 'sansserif':29, 'serif':30}
    
    _font_family = [['arial', 'helvetica', 'liberation sans',  'nimbus sans l', 'freesans', 'tahoma', 'sans-serif'], ['verdana', 'bitstream vera sans', 'dejavu sans', 'sans-serif'], ['impact', 'sans-serif'], ['comic sans ms', 'cursive', 'sans-serif'], ['courier new', 'courier', 'lucida console', 'dejavu sans mono', 'monospace'], ['times new roman', 'times', 'liberation serif', 'nimbus roman no9 l', 'serif'], ['garamond',  'book antiqua', 'palatino', 'liberation serif', 'nimbus roman no9 l', 'serif'], ['georgia', 'bitstream vera serif', 'lucida serif', 'liberation serif', 'dejavu serif', 'serif']]

    def __init__(self, name, size):
        """
        Return Font object.
        Arguments include name of a system font and size of font. The name argument can be a string of comma-delimited names to specify fallbacks and use a default font if none found.
        """
        if not name:
            name = Font._font[0]
        font = [fn.strip().lower() for fn in name.split(',')]
        fallback = None
        for i, fn in enumerate(font):
            if fn in Font._font:
                if not fallback:
                    fallback = fn
            else:
                f = ''.join(c for c in fn if c.isalnum())
                if f in Font._font_alt:
                    font[i] = Font._font[Font._font_alt[f]]
                    if not fallback:
                        fallback = font[i]
        if fallback:
            for ff in Font._font_family:
                if fallback in ff:
                    font.extend(f for f in ff if f not in font)
                    break
        else:
            font.extend(Font._font_family[0])
        self.fontname = ','.join(font)
        self.fontsize = size
        self.bold = ''
        self.italic = ''
        self.fontstyle = self.bold + ' ' + self.italic
        self.underline = False
        self.char_size = None
        self._nonimplemented_methods()

    def __repr__(self):
        """
        Return string representation of Font object.
        """
        return "%s(%r)" % (self.__class__, self.__dict__)

    def render(self, text, antialias=True, color=(0,0,0), background=None, surface=None):      #optional surface for text rendering
        """
        Render text onto surface.
        Arguments are text to render, and optional antialias, RGB color of text, RGB color of background, and surface for text rendering.
        """
        if not surface:
            w,h = self.size(text)
            surf = Surface((w,h))
        else:
            surf = surface
            w,h = surface.width, surface.height
        if background:
            surf.setFillStyle(Color(background))
            surf.fillRect(0,0,w,h)
        surf.setFont('%s %dpx %s' % (self.fontstyle, self.fontsize, self.fontname))
#        if antialias: pass
        surf.setFillStyle(Color(color))
        surf.fillText(text,0,self.fontsize)
        if self.underline:
            surf.setLineWidth(1)
            surf.setStrokeStyle(Color(color))
            surf.setStroke(BasicStroke(1))
            surf.moveTo(0, h-1)
            surf.lineTo(w-1, h-1)
            surf.stroke()
        return surf

    def size(self, text):
        """
        Return size x,y of a surface for of given text.
        """
        if _surf:   #>IE9 - use exception if HTML5Canvas not implemented
            _surf.setFont('%s %dpx %s' % (self.fontstyle, self.fontsize, self.fontname))
            x = _surf.measureText(text)
        else:   #estimate
            x = self._size_estimate(text)
        if x < 1:
            x = 1
        y = self.fontsize + 5
        return (x, y)

    def _size_estimate(self, text=None):   #for browsers HTML5Canvas not implemented
        if not self.char_size:
            self.char_size = self._get_char_size()
        self.fontname = ','.join(Font._font_family[0])
        self.fontstyle = ''
        size = []
        for char in text:
            try:
                size.append(self.char_size[char] * self.fontsize)
            except KeyError:
                size.append(self.char_size['x'] * self.fontsize)
        x = math.ceil( sum(size) )
        return x

    def set_underline(self, setting=True):
        """
        Set font underline style.
        Optional setting argument, default to True.
        """
        self.underline = setting

    def get_underline(self):
        """
        Check if font is underlined.
        """
        return self.underline

    def set_bold(self, setting=True):
        """
        Set font bold style.
        Optional setting argument, default to True.
        """
        self.bold = {True:'bold', False:''}[setting]
        self.fontstyle = self.bold + ' ' + self.italic

    def get_bold(self):
        """
        Check if font is bold.
        """
        if self.bold:
            return True
        else:
            return False

    def set_italic(self, setting=True):
        """
        Set font italic style.
        Optional setting argument, default to True.
        """
        self.italic = {True:'italic', False:''}[setting]
        self.fontstyle = self.bold + ' ' + self.italic

    def get_italic(self):
        """
        Check if font is italized.
        """
        if self.italic:
            return True
        else:
            return False

    def get_linesize(self):
        """
        Return linesize of font.
        """
        return int(self.fontsize*1.2)

    def _nonimplemented_methods(self):
        """
        Non-implemented methods.
        """
        self.metrics = lambda *arg: []
        self.get_height = lambda *arg: 0
        self.get_ascent = lambda *arg: 0
        self.get_descent = lambda *arg: 0

    def _get_char_size(self, font=None):    #for browsers HTML5Canvas not implemented
        if not font:
            return {'a': 0.6, 'b': 0.6, 'c': 0.5, 'd': 0.6, 'e': 0.6, 'f': 0.3, 'g': 0.6, 'h': 0.6, 'i': 0.2, 'j': 0.2, 'k': 0.5, 'l': 0.2, 'm': 0.8, 'n': 0.6, 'o': 0.6, 'p': 0.6, 'q': 0.6, 'r': 0.3, 's': 0.5, 't': 0.3, 'u': 0.6, 'v': 0.5, 'w': 0.7, 'x': 0.5, 'y': 0.5, 'z': 0.5, 'A': 0.7, 'B': 0.7, 'C': 0.7, 'D': 0.7, 'E': 0.7, 'F': 0.6, 'G': 0.8, 'H': 0.7, 'I': 0.3, 'J': 0.5, 'K': 0.7, 'L': 0.6, 'M': 0.8, 'N': 0.7, 'O': 0.8, 'P': 0.7, 'Q': 0.8, 'R': 0.7, 'S': 0.7, 'T': 0.6, 'U': 0.7, 'V': 0.7, 'W': 0.9, 'X': 0.7, 'Y': 0.7, 'Z': 0.6, '0': 0.6, '1': 0.6, '2': 0.6, '3': 0.6, '4': 0.6, '5': 0.6, '6': 0.6, '7': 0.6, '8': 0.6, '9': 0.6, '.': 0.3, ',': 0.3, ':': 0.3, ';': 0.3, '?': 0.6, '~': 0.6, '!': 0.3, '@': 1, '#': 0.6, '$': 0.6, '%': 0.9, '^': 0.5, '&': 0.7, '=': 0.6, '+': 0.6, '-': 0.3, '*': 0.4, '/': 0.3, '\\': 0.3, '_': 0.6, '<': 0.6, '>': 0.6, '(': 0.3, ')': 0.3, '{': 0.3, '}': 0.3, '[': 0.3, ']': 0.3, "'": 0.2, '"': 0.4, ' ': 0.3}
        else:
            fontsize = 10
            _surf.setFont('%dpx %s' % (fontsize, font))     #generated font='arial'
            char_size = {}
            for char in 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.,:;?~!@#$%^&=+-*/\_<>(){}[]\'\" ':
                char_size[char] = float(_surf.measureText(char)/fontsize)
            return char_size


class SysFont(Font):
    """
    **pyjsdl.font.SysFont**
    
    * Font subclass
    """

    def __init__(self, name, size, bold=False, italic=False):
        """
        Return SysFont subclassed of Font.
        Arguments include name of a system font and size of font, with optional bold and italic style. The name argument can be a string of comma-delimited names to specify fallbacks and use a default font if none found.
        """
        Font.__init__(self,name,size)
        self.bold = {True:'bold', False:''}[bold]
        self.italic = {True:'italic', False:''}[bold]
        self.fontstyle = self.bold + ' ' + self.italic

